src/core/ComboBox/ComboBox.tsx

```diff
@@ -117,252 +122,312 @@ export const ComboBox = <T,>(props: ComboBoxProps<T>) => {
+      const customItem = itemRenderer
+        ? itemRenderer(option, {
+            isFocused: false,
+            isSelected: false,
+            index: __originalIndex,
+            id: option.id,
+          })
+        : null;
+
+      return customItem ? (
+        React.cloneElement(customItem, {
+          'data-iui-index': __originalIndex,
+          onClick: (e: unknown) => {
+            dispatch(['select', __originalIndex as number]);
+            customItem.props.onClick?.(e);
+          },
+        })
+      ) : (
+        <ComboBoxMenuItem
+          key={(__originalIndex as number) ?? index}
+          data-iui-index={option.__originalIndex}
+          onClick={() => dispatch(['select', __originalIndex as number])}
+          {...rest}
+        >
+          {option.label}
+        </ComboBoxMenuItem>
       );
-    setFilteredOptions(_filteredOptions);
-
-    setFocusedIndex((previouslyFocusedIndex) => {
-      if (_filteredOptions.includes(options[previouslyFocusedIndex])) {
-        return previouslyFocusedIndex;
-      } else if (
-        _filteredOptions.find(({ value }) => value === selectedValue)
-      ) {
-        return options.findIndex(({ value }) => value === selectedValue);
-      } else {
-        return -1; // reset focus if previously focused or selected value is not in filtered list
-      }
     });
-  }, [inputValue, options, selectedValue, isOpen, filterFunction]);
-
-  const onKeyDown = React.useCallback(
-    (event: React.KeyboardEvent) => {
-      const focusableOptions = getFocusableElements(menuRef.current);
-      const focusedIndexInFilteredList = focusableOptions.findIndex(
-        ({ id }) =>
-          id === inputRef.current?.getAttribute('aria-activedescendant'),
+  }, [itemRenderer, options]);
+
+  const getMenuItem = React.useCallback(
+    (option: ComboBoxOption<T>) => {
+      const { __originalIndex, ...rest } = option;
+
+      const customItem = itemRenderer
+        ? itemRenderer(option, {
+            isFocused: focusedIndex === __originalIndex,
+            isSelected: selectedIndex === __originalIndex,
+            index: __originalIndex,
+            id: option.id,
+          })
+        : null;
+
+      return selectedIndex === __originalIndex ||
+        focusedIndex === __originalIndex ? (
+        customItem ? (
+          React.cloneElement(customItem, {
+            className: cx(customItem.props.className, 'iui-focused'),
+            'data-iui-index': __originalIndex,
+            onClick: (e: unknown) => {
+              dispatch(['select', __originalIndex as number]);
+              customItem.props.onClick?.(e);
+            },
+          })
+        ) : (
+          <ComboBoxMenuItem
+            key={option.id ?? __originalIndex}
+            {...rest}
+            isSelected={selectedIndex === __originalIndex}
+            isFocused={focusedIndex === __originalIndex}
+            onClick={() => dispatch(['select', __originalIndex])}
```

Comment:
Click is still laggy.

Reply:
Doesn't lag when virtualized. I'm not sure how to fix it in regular mode when there's a large number of items. In my `dispatch(['select'])` I'm setting `isOpen: false` but that triggers a whole bunch of updates before the popover actually closes.

Reply:
I think the problem here is that you change `selectedIndex` and `value` is changed from outside that causes full rerender. But if with virtual scroll it is working fine, maybe no need to dig deeper.